Firestore/core/src/local/leveldb_migrations.cc

```diff
@@ -303,6 +307,76 @@ void EnsureCollectionParentsIndex(leveldb::DB* db) {
   transaction.Commit();
 }
 
+/**
+ * Returns a `TargetData` by reading the `targets` table, using the given key
+ * for `query_targets` as a foreign key.
+ */
+util::StatusOr<TargetData> ReadTargetData(
+    const LevelDbQueryTargetKey& query_target_key,
+    const LocalSerializer& serializer,
+    LevelDbTransaction& transaction) {
+  auto target_it = transaction.NewIterator();
+  const auto& target_key = LevelDbTargetKey::Key(query_target_key.target_id());
+  target_it->Seek(target_key);
+  if (!target_it->Valid()) {
+    LOG_WARN("Dangling query-target reference found: seeking %s found %s",
+             DescribeKey(target_key), DescribeKey(target_it));
+    return util::Status(kErrorNotFound,
+                        "Dangling query-target reference found.");
+  }
+
+  StringReader reader{target_it->value()};
+  auto message = Message<firestore_client_Target>::TryParse(&reader);
+  if (!reader.ok()) {
+    HARD_FAIL("Target proto failed to parse: %s", reader.status().ToString());
```

Comment:
Can we return a failed status here (and below)? It is probably best not to crash during schema conversion since we load all existing data.

Reply:
Done.